### PR TITLE
Added limit on the size of messages queue in bytes.

### PR DESCRIPTION
**Pull request by vadim-kovalyov**
This is continuation of work started in #3063. In this PR:
- Added limit of the messages queue by size (payload size).
- Refactored queue limit logic into `BoundedQueue` struct.
- Addressed some feedback from previous PR.

### CODE REVIEW

File: mqtt/mqtt-broker/src/session.rs [hunk 1]
```diff
@@ -910,6 +907,103 @@ impl Default for PacketIdentifiers {
...
+        BoundedQueue {
+            inner: VecDeque::new(),
+            max_len,
+            max_size,
+            when_full,
+            cur_size: 0,
+        }
+    }
+
+    pub fn from_iter(
```
[the remark this answers is not in the document]
Comment (dmolokanov) [reply mqtt/mqtt-broker/src/session.rs:940]: some ideas: 
1. remove this method and move the part that moves items into queue to https://doc.rust-lang.org/std/iter/trait.Extend.html implementation.
2. rename `from_iter` into something with a more clear name

File: mqtt/mqtt-broker/src/session.rs [hunk 2]
```diff
@@ -910,6 +907,103 @@ impl Default for PacketIdentifiers {
...
 }
 
+/// `BoundedQueue` is a queue of publications with bounds by count and total payload size in bytes.
+///
+/// Packets will be queued until either of `max_len` or `max_size` limits is reached, and
+/// then `when_full` strategy is applied.
+///
+/// None for `max_len` or `max_size` means "unbounded".
+#[derive(Clone, Debug, PartialEq)]
+struct BoundedQueue {
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/session.rs:917]: I think this code deserves separate module and more tests in isolation to cover corner cases

File: mqtt/mqtt-broker/src/session.rs [hunk 3]
```diff
@@ -910,6 +907,103 @@ impl Default for PacketIdentifiers {
...
+    inner: VecDeque<proto::Publication>,
+    max_len: Option<NonZeroUsize>,
+    max_size: Option<NonZeroU64>,
+    when_full: QueueFullAction,
+    cur_size: u64,
```
[the remark this answers is not in the document]
Comment (vadim-kovalyov) [reply mqtt/mqtt-broker/src/session.rs:922]: I can put a code comment for that. I'm not sure what else. Any suggestions?

File: mqtt/mqtt-broker/src/session.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (vadim-kovalyov) [reply mqtt/mqtt-broker/src/session.rs:940]: Makes sense. I'll try Extend trait.

File: mqtt/mqtt-broker/src/session.rs (hunk 2 above)
Comment (vadim-kovalyov) [reply mqtt/mqtt-broker/src/session.rs:919]: Maybe, but I was not trying to make a publicly available general purpose collection, rather just refactor some code a bit. I would keep it as internal details of `SessionState` and would rather move SessionState+BoundedQueue into a separate module.

File: mqtt/mqtt-broker/src/configuration.rs [hunk 4]
```diff
@@ -77,16 +80,20 @@ impl SessionConfig {
         }
     }
 
-    pub fn max_inflight_messages(&self) -> usize {
-        self.max_inflight_messages
+    pub fn max_inflight_messages(&self) -> Option<NonZeroUsize> {
+        NonZeroUsize::new(self.max_inflight_messages)
```
[the remark this answers is not in the document]
Comment (vadim-kovalyov) [reply mqtt/mqtt-broker/src/configuration.rs:84]: It looks like it does not handle in the way we need. We can't just define field as `Option<NonZeroUsize>` since it uses Option<T> to denote presence of the field, not that it has invalid value in the json. As an alternative - create newtype and implement custom deserializer. But IMHO it is unnecessary here.

File: mqtt/mqtt-broker/src/session.rs [hunk 5]
```diff
@@ -910,6 +907,103 @@ impl Default for PacketIdentifiers {
...
+}
+
+impl BoundedQueue {
+    pub fn new(
+        max_len: Option<NonZeroUsize>,
+        max_size: Option<NonZeroU64>,
+        when_full: QueueFullAction,
```
[the remark this answers is not in the document]
Comment (vadim-kovalyov) [reply mqtt/mqtt-broker/src/session.rs:932]: I see your point, but it's a minor detail of a private API. I was not trying to make a publicly available general purpose collection, rather just refactor some code a bit. I'm not sure that polishing API worth the effort at the moment.

File: mqtt/mqtt-broker/src/configuration.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmolokanov) [reply mqtt/mqtt-broker/src/configuration.rs:84]: Agreed. Thanks for explanation!

File: mqtt/mqtt-broker/src/session.rs (hunk 2 above)
Comment (dmolokanov) [reply mqtt/mqtt-broker/src/session.rs:920]: Do you think we have enough tests to cover queue limits?

File: mqtt/mqtt-broker/src/session.rs [hunk 6]
```diff
@@ -910,6 +909,99 @@ impl Default for PacketIdentifiers {
...
+            if self.current_size + pub_len > max_size.get() {
+                return self.handle_queue_limit(publication);
+            }
+        }
+
+        self.current_size += publication.payload.len() as u64;
+        self.inner.push_back(publication);
+    }
+
+    #[cfg(test)]
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/session.rs:975]: nit: should we make it available for everything?

File: mqtt/mqtt-broker/src/session.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmolokanov) [reply mqtt/mqtt-broker/src/session.rs:932]: makes sense. rust provides reach type system, we should employ it though

File: mqtt/mqtt-broker/src/session.rs (hunk 6 above)
Comment (vadim-kovalyov) [reply mqtt/mqtt-broker/src/session.rs:977]: So far it just used in tests, otherwise give dead code warning. We can remove `cfg` when we need to use this in prod code.

File: mqtt/mqtt-broker/src/session.rs (hunk 2 above)
Comment (vadim-kovalyov) [reply mqtt/mqtt-broker/src/session.rs:922]: I think I've covered all cases I could think of, but they cover `publish_to` public method that in turn uses private `BoundedQueue`.
